planner/core/indexmerge_path.go

```diff
@@ -642,16 +628,8 @@ func (ds *DataSource) generateIndexMergeJSONMVIndexPath(normalPathCnt int, filte
 }
 
 // jsonArrayExpr2Exprs converts a JsonArray expression to expression list: cast('[1, 2, 3]' as JSON) --> []expr{1, 2, 3}
-func jsonArrayExpr2Exprs(sctx sessionctx.Context, jsonArrayExpr expression.Expression) ([]expression.Expression, bool) {
-	// only support cast(const as JSON)
-	arrayExpr, wrappedByJSONCast := unwrapJSONCast(jsonArrayExpr)
-	if !wrappedByJSONCast {
-		return nil, false
-	}
-	if _, isConst := arrayExpr.(*expression.Constant); !isConst {
-		return nil, false
-	}
-	if expression.IsMutableEffectsExpr(arrayExpr) {
+func jsonArrayExpr2Exprs(sctx sessionctx.Context, jsonArrayExpr expression.Expression, targetType *types.FieldType) ([]expression.Expression, bool) {
+	if !expression.IsInmutableExpr(jsonArrayExpr) || jsonArrayExpr.GetType().EvalType() != types.ETJson {
```

Comment:
how about use `FoldConst`?

Reply:
No, I've tried this and found that `FoldConst` cannot handle expressions with nested Scalar-Functions like `add(1, add(1, 1))`.